src/sources/audiosourcemp3.cpp

```diff
@@ -0,0 +1,545 @@
+            // out of range -> abort
+            return m_curFrameIndex;
+        }
+        DEBUG_ASSERT(findSeekFrameIndex(m_curFrameIndex) == seekFrameIndex);
+    }
+
+    // Decoding starts before the actual target position
+    DEBUG_ASSERT(m_curFrameIndex <= frameIndex);
+
+    // Skip (= decode and discard) prefetch data
+    const size_type skipFrameCount = frameIndex - m_curFrameIndex;
+    skipFrameSamples(skipFrameCount);
+    DEBUG_ASSERT(m_curFrameIndex == frameIndex);
+
+    DEBUG_ASSERT(isValidFrameIndex(m_curFrameIndex));
+    return m_curFrameIndex;
+}
+
+AudioSource::size_type AudioSourceMp3::readSampleFrames(
+        size_type numberOfFrames, sample_type* sampleBuffer) {
+    return readSampleFrames(numberOfFrames,
+            sampleBuffer, frames2samples(numberOfFrames),
+            false);
+}
+
+AudioSource::size_type AudioSourceMp3::readSampleFramesStereo(
+        size_type numberOfFrames, sample_type* sampleBuffer,
+        size_type sampleBufferSize) {
+    return readSampleFrames(numberOfFrames,
+            sampleBuffer, sampleBufferSize,
+            true);
+}
+
+AudioSource::size_type AudioSourceMp3::readSampleFrames(
+        size_type numberOfFrames, sample_type* sampleBuffer,
+        size_type sampleBufferSize, bool readStereoSamples) {
+    DEBUG_ASSERT(isValidFrameIndex(m_curFrameIndex));
+    DEBUG_ASSERT(getSampleBufferSize(numberOfFrames, readStereoSamples) <= sampleBufferSize);
+
+    const size_type numberOfFramesTotal = math_min(numberOfFrames,
+            size_type(getFrameIndexMax() - m_curFrameIndex));
+
+    sample_type* pSampleBuffer = sampleBuffer;
+    size_type numberOfFramesRemaining = numberOfFramesTotal;
+    while (0 < numberOfFramesRemaining) {
+        if (0 >= m_madSynthCount) {
+            // When all decoded output data has been consumed...
+            DEBUG_ASSERT(0 == m_madSynthCount);
+            // ...decode the next MP3 frame
+            DEBUG_ASSERT(NULL != m_madStream.buffer);
+            DEBUG_ASSERT(NULL != m_madStream.this_frame);
+
+            // WARNING: Correctly evaluating and handling the result
+            // of mad_frame_decode() has proven to be extremely tricky.
+            // Don't change anything at the following lines of code
+            // unless you know what you are doing!!!
+            unsigned char const* pMadThisFrame = m_madStream.this_frame;
+            if (0 != mad_frame_decode(&m_madFrame, &m_madStream)) {
+                if (MAD_RECOVERABLE(m_madStream.error)) {
+                    if (pMadThisFrame != m_madStream.this_frame) {
+                        // Suppress common cases of "lost synchronization"
+                        // warnings. When seeking through the file those
+                        // warnings are expected while skipping over
+                        // prefetched frames.
+                        if ((MAD_ERROR_LOSTSYNC == m_madStream.error) && (NULL == pSampleBuffer)) {
```

Comment:
NULL == pSampleBuffer means that we will discard the decoded samples immediately (prefetch after seek). I though about mentioning this in the comment. Well, I better add this as a remark ;)